### ISSUE TITLE
[Neo4j Crashes] neostore.propertystore.db.strings.id potentially corrupted

### ISSUE THREAD

**Issue by arun-bedrock** [neo4j/neo4j#13538]
Neo4j is crashing after some indeterminate amount of time after small batch writes and individual reads.  This error occurs has occurred randomly when running batch writes of about 50 edges (20 concurrent writes) so 1000 edge writes at a time. It sometimes happens after 10 minutes and sometimes after 3 hours of batch writing. Reads and writes are individually quite fast (less than 1-2 seconds in the worst case and generally less than 100ms in most cases). We get this error that shows up on the neo4j and it crashes the instance and we have to restart the neo4j. 

**Neo4j Version:** 4.4
**Operating System:** Linux/UNIX. EC2 m5.24xlarge 
**API:** Golang Driver 5.0

### Steps to reproduce

We run a cheap read query and then split up the writes to be a maximum of 50 edges (in case we get more edges from the read).

The batch writes look like this with a size of between 500 KB to 3 MB:
```
UNWIND $items AS item		
MATCH (n1:Principal {key: item.fromKey})
MATCH (n2:Datastore {key: item.toKey})
MERGE (n1)-[r1:ENTITLED]->(n2)
SET r1 += item.properties
RETURN r1
```
Eventually, the writes and reads slow down dramatically and then the database crashes and needs to be restarted. I'm attaching the stack trace and our neo4j conf if we're set something up incorrectly. 
[debug (4).log](https://github.com/user-attachments/files/17399685/debug.4.log)
[neo4j.conf.txt](https://github.com/user-attachments/files/17399719/neo4j.conf.txt)

### Expected behavior
Neo4j should not fail in a case like this. 

### Actual behavior
Neo4j crashes with this stack trace:
```
2024-10-15 23:01:26.470+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.Transaction.TransactionLogError]: Could not append transaction representation to log, reference 58c2a7a1-d54b-4cfd-a82a-143dd31c1cd9.

Caused by: org.neo4j.internal.kernel.api.exceptions.TransactionApplyKernelException: Failed to apply transaction: Transaction #488506843 at log position LogPosition{logVersion=18367, byteOffset=201482813} {started 2024-10-15 23:01:26.431+0000, committed 2024-10-15 23:01:26.439+0000, with 52 commands in this transaction, lease -1, latest committed transaction id when started was 488506829, additional header bytes: }

Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Illegal addition ID state for range: 26760914 transition longIdx: 1
into: 00000000 00000000 00000000 00000000 00000000 00000000 00111110 00000010
from: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000010 | GB+Tree[file:/var/lib/neo4j/data/databases/neo4j/[neostore.propertystore.db.strings.id](http://neostore.propertystore.db.strings.id/), layout:IdRangeLayout[version:1.2, identifier:3735929182, fixedSize:true], generation:14677/14678]
```

I saved this index file as well. We made a copy of the neo4j instance a few days ago and one of our most recent runs succeeded on the copy but failed on the original instance. We're not sure why this happened -- potentially this `neostore.propertystore.db.strings.id` is corrupted.

**Comment by arun-bedrock** [comment]
Hi, wanted to follow up on this!

**Comment by arun-bedrock** [comment]
Hi, we've encountered new kernel issues with the same application running. I stopped the old neo4j and deleted the `neostore.propertystore.db.strings.id` and neo4j recreated it successfully to fix the above issue. I don't know how to fix the below issue. 

We will likely have to switch off neo4j because of these issues if there's not a resolution. Would appreciate some support. 

Here is the new issue's stack trace:
```
2024-10-24 22:25:14.406+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.Transaction.TransactionCommitFailed]: Unable to complete transaction., reference 98c7ee7b-5b76-4002-9fc7-ae21e0c403d1.
org.neo4j.fabric.executor.FabricException: Unable to complete transaction.
	at org.neo4j.fabric.executor.Exceptions.transform(Exceptions.java:65) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.Exceptions.transform(Exceptions.java:35) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.throwIfNonEmpty(FabricTransactionImpl.java:317) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.commit(FabricTransactionImpl.java:222) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.bolt.BoltFabricDatabaseService$BoltTransactionImpl.commit(BoltFabricDatabaseService.java:146) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.closeTransaction(TransactionStateMachine.java:404) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$2.commitTransaction(TransactionStateMachine.java:356) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.commitTransaction(TransactionStateMachine.java:108) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.transaction.StatementProcessorTxManager.commit(StatementProcessorTxManager.java:66) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processCommitMessage(InTransactionState.java:107) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v44.runtime.InTransactionState.processCommitMessage(InTransactionState.java:35) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processUnsafe(InTransactionState.java:56) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:150) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:98) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:93) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:156) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:250) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:187) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:177) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:257) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$3(ExecutorBoltScheduler.java:240) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.neo4j.graphdb.TransactionFailureException: Unable to complete transaction.
	at org.neo4j.fabric.executor.FabricLocalExecutor$LocalTransactionContext.transformTerminalOperationError(FabricLocalExecutor.java:237) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.safeTerminalOperation(TransactionImpl.java:708) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:174) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:169) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricKernelTransaction.commit(FabricKernelTransaction.java:137) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricLocalExecutor$KernelTxWrapper.doCommit(FabricLocalExecutor.java:269) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:73) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:32) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:127) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:201) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:193) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:258) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:78) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.block(Mono.java:1706) ~[reactor-core-3.4.15.jar:3.4.15]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.doOnChildren(FabricTransactionImpl.java:296) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.commit(FabricTransactionImpl.java:207) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	... 23 more
Caused by: org.neo4j.kernel.impl.store.InvalidRecordException: Property[1018859289,used=false,prev=1019313624,next=1018859288, (blocks not loaded)] not in use
	at org.neo4j.kernel.impl.store.record.RecordLoad.verify(RecordLoad.java:141) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.store.CommonAbstractStore.verifyAfterReading(CommonAbstractStore.java:1074) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.store.CommonAbstractStore.readRecordFromPage(CommonAbstractStore.java:897) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.store.CommonAbstractStore.readIntoRecord(CommonAbstractStore.java:850) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.store.CommonAbstractStore.getRecordByCursor(CommonAbstractStore.java:830) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.Loaders$RecordLoader.load(Loaders.java:394) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.Loaders$2.load(Loaders.java:226) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.Loaders$2.load(Loaders.java:204) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordChanges.getOrLoad(RecordChanges.java:96) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordAccess.getOrLoad(RecordAccess.java:37) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.PropertyCreator.primitiveSetProperty(PropertyCreator.java:88) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.TransactionRecordState.relAddProperty(TransactionRecordState.java:439) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.TransactionToRecordStateVisitor.visitAddedRelProperties(TransactionToRecordStateVisitor.java:131) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.TransactionToRecordStateVisitor.visitRelPropertyChanges(TransactionToRecordStateVisitor.java:124) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.api.txstate.TxStateVisitor$Delegator.visitRelPropertyChanges(TxStateVisitor.java:190) ~[neo4j-storage-engine-api-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.state.TxState.lambda$accept$0(TxState.java:217) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.state.RelationshipStateImpl.accept(RelationshipStateImpl.java:138) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.state.TxState.accept(TxState.java:216) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordStorageEngine.createCommands(RecordStorageEngine.java:397) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commitTransaction(KernelTransactionImplementation.java:857) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.closeTransaction(KernelTransactionImplementation.java:744) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commit(KernelTransactionImplementation.java:715) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.lambda$commit$0(TransactionImpl.java:179) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.safeTerminalOperation(TransactionImpl.java:699) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:174) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:169) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricKernelTransaction.commit(FabricKernelTransaction.java:137) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricLocalExecutor$KernelTxWrapper.doCommit(FabricLocalExecutor.java:269) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:73) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:32) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:127) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:201) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:193) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:258) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:78) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.block(Mono.java:1706) ~[reactor-core-3.4.15.jar:3.4.15]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.doOnChildren(FabricTransactionImpl.java:296) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.commit(FabricTransactionImpl.java:207) ~[neo4j-fabric-4.4.11.jar:4.4.11]
```

**Comment by tinwelint** [comment]
Hi @arun-bedrock the initial issue, e.g:

````
into: 00000000 00000000 00000000 00000000 00000000 00000000 00111110 00000010
from: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000010 | GB+Tree[file:/var/lib/neo4j/data/databases/neo4j/[neostore.propertystore.db.strings.id](http://neostore.propertystore.db.strings.id/), layout:IdRangeLayout[version:1.2, identifier:3735929182, fixedSize:true], generation:14677/14678]
````

looks like there's a record ID (in this case a record holding some string property value) that Neo4j wants to mark as deleted, but it's already deleted. Why this happens I don't know. The newer `InvalidRecordException` is almost certainly a direct consequence of the bug causing the former issue.

Before investigating too much into your issue: is there some requirement that you run on the very old 4.4.11 version? I think you should either try to run the latest 4.4, or ideally move to the latest 5.x. The debug log says you're running enterprise edition, which makes this ideal if you move to 5.x because you can migrate to the new `block` storage format instead. See https://neo4j.com/docs/operations-manual/current/database-internals/store-formats/

**Comment by arun-bedrock** [comment]
Hi @tinwelint, we updated Neo4j to 4.4.38 and it had the same issue. We tested deleting the edges and recreating them and that fixed this issue. We can investigate doing a migration to 5.0 but that'd be a big undertaking. 

Do you know how we can avoid getting into this state to begin with?

**Comment by tinwelint** [comment]
@arun-bedrock Unfortunately there are no known issues causing an issue like this. To help out finding this issue you could set the config:

````
unsupported.dbms.idgenerator.log.enabled=true
````

which would log internal ID events into e.g. `neostore.propertystore.db.strings.id.log` (w/ or w/o some timestamp after it). Running into the issue w/ this config set _could_ help me shed some light on some ordering of events causing this. This file contains no domain data of any sort. Worst case I'd also need the transaction log, if that would be OK for you to share.

Would this be something to pursuit, you think?

**Comment by arun-bedrock** [comment]
@tinwelint would setting this log statement have any performance implications? We wouldn't be able to share the transaction log because I have no way to separate out the customer information.

**Comment by tinwelint** [comment]
@arun-bedrock it has _some_ overhead, although on the system as a whole it should be quite small.

**Comment by arun-bedrock** [comment]
Ok we're still experiencing this issue -- I added that conf line `unsupported.dbms.idgenerator.log.enabled=true`. I will post the output the next time it crashes

**Comment by arun-bedrock** [comment]
Hi here is a zip of the output with those id logs (125 mb zip -> 861 mbs uncompressed):
https://drive.google.com/file/d/11fPC1jaWLAtk8vXJXbU9mXr8Rg4UW-xQ/view?usp=sharing

I attached the debug.log.1 for this crash in the zip as well. This is a preview of the crash we saw from the debug.log.1 file: 
```
2024-11-11 02:37:57.049+0000 INFO  [o.n.k.d.Database] [neo4j/6d79d176] Rotated to transaction log [/var/lib/neo4j/data/transactions/neo4j/neostore.transaction.db.22239] version=22238, last transaction in previous log=558712014, rotation took 30 millis, started after 69810 millis.
2024-11-11 02:38:30.564+0000 ERROR [o.n.m.DatabaseHealth] [neo4j/6d79d176] Database panic: The database has encountered a critical error, and needs to be restarted. Please see database logs for more details.
org.neo4j.internal.kernel.api.exceptions.TransactionApplyKernelException: Failed to apply transaction: Transaction #558712032 at log position LogPosition{logVersion=22239, byteOffset=140812923} {started 2024-11-11 02:38:28.739+0000, committed 2024-11-11 02:38:30.453+0000, with 17056 commands in this transaction, lease -1, latest committed transaction id when started was 558712031, additional header bytes: }
	at org.neo4j.internal.recordstorage.RecordStorageEngine.apply(RecordStorageEngine.java:515) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.InternalTransactionCommitProcess.applyToStore(InternalTransactionCommitProcess.java:81) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.InternalTransactionCommitProcess.commit(InternalTransactionCommitProcess.java:53) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.DatabaseTransactionCommitProcess.commit(DatabaseTransactionCommitProcess.java:42) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commitTransaction(KernelTransactionImplementation.java:892) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.closeTransaction(KernelTransactionImplementation.java:744) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commit(KernelTransactionImplementation.java:715) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.lambda$commit$0(TransactionImpl.java:179) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.safeTerminalOperation(TransactionImpl.java:699) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:174) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.commit(TransactionImpl.java:169) ~[neo4j-kernel-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricKernelTransaction.commit(FabricKernelTransaction.java:137) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.executor.FabricLocalExecutor$KernelTxWrapper.doCommit(FabricLocalExecutor.java:269) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:73) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoRunnable.call(MonoRunnable.java:32) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:127) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:201) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2398) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:193) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:258) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:78) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.15.jar:3.4.15]
	at reactor.core.publisher.Mono.block(Mono.java:1706) ~[reactor-core-3.4.15.jar:3.4.15]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.doOnChildren(FabricTransactionImpl.java:296) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.transaction.FabricTransactionImpl.commit(FabricTransactionImpl.java:207) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.fabric.bolt.BoltFabricDatabaseService$BoltTransactionImpl.commit(BoltFabricDatabaseService.java:146) ~[neo4j-fabric-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.closeTransaction(TransactionStateMachine.java:404) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$2.commitTransaction(TransactionStateMachine.java:356) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.commitTransaction(TransactionStateMachine.java:108) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.transaction.StatementProcessorTxManager.commit(StatementProcessorTxManager.java:66) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processCommitMessage(InTransactionState.java:107) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v44.runtime.InTransactionState.processCommitMessage(InTransactionState.java:35) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processUnsafe(InTransactionState.java:56) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:150) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:98) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:93) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:156) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:250) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:187) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:177) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:257) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$3(ExecutorBoltScheduler.java:240) ~[neo4j-bolt-4.4.11.jar:4.4.11]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.io.IOException: Exception closing multiple resources.
	at org.neo4j.io.IOUtils.close(IOUtils.java:159) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.close(IOUtils.java:188) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.closeAll(IOUtils.java:116) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.BatchContextImpl.close(BatchContextImpl.java:79) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordStorageEngine.apply(RecordStorageEngine.java:511) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	... 52 more
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Illegal addition ID state for range: 8795398 transition longIdx: 1
into: 00001111 11000000 00000011 11111111 11000000 00000000 00001100 01111000
from: 00000000 01000000 00000000 00000000 00000000 00000000 00000000 00000000 | GB+Tree[file:/var/lib/neo4j/data/databases/neo4j/neostore.propertystore.db.strings.id, layout:IdRangeLayout[version:1.2, identifier:3735929182, fixedSize:true], generation:9/10]
	at org.neo4j.util.concurrent.WorkSync.checkFailure(WorkSync.java:178) ~[neo4j-concurrent-4.4.11.jar:4.4.11]
	at org.neo4j.util.concurrent.WorkSync$1.await(WorkSync.java:130) ~[neo4j-concurrent-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$ChangedIds.awaitApply(IdGeneratorUpdatesWorkSync.java:173) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.awaitApply(IdGeneratorUpdatesWorkSync.java:113) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.apply(IdGeneratorUpdatesWorkSync.java:103) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.close(IdGeneratorUpdatesWorkSync.java:129) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.close(IOUtils.java:152) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.close(IOUtils.java:188) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.closeAll(IOUtils.java:116) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.BatchContextImpl.close(BatchContextImpl.java:79) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordStorageEngine.apply(RecordStorageEngine.java:511) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	... 52 more
Caused by: java.lang.IllegalStateException: Illegal addition ID state for range: 8795398 transition longIdx: 1
into: 00001111 11000000 00000011 11111111 11000000 00000000 00001100 01111000
from: 00000000 01000000 00000000 00000000 00000000 00000000 00000000 00000000 | GB+Tree[file:/var/lib/neo4j/data/databases/neo4j/neostore.propertystore.db.strings.id, layout:IdRangeLayout[version:1.2, identifier:3735929182, fixedSize:true], generation:9/10]
	at org.neo4j.internal.id.indexed.IdRange.verifyMerge(IdRange.java:236) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.internal.id.indexed.IdRange.mergeFrom(IdRange.java:148) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.internal.id.indexed.IdRangeMerger.merge(IdRangeMerger.java:57) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.internal.id.indexed.IdRangeMerger.merge(IdRangeMerger.java:33) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.index.internal.gbptree.InternalTreeLogic.mergeValue(InternalTreeLogic.java:571) ~[neo4j-index-4.4.11.jar:4.4.11]
	at org.neo4j.index.internal.gbptree.InternalTreeLogic.insertInLeaf(InternalTreeLogic.java:554) ~[neo4j-index-4.4.11.jar:4.4.11]
	at org.neo4j.index.internal.gbptree.InternalTreeLogic.insert(InternalTreeLogic.java:398) ~[neo4j-index-4.4.11.jar:4.4.11]
	at org.neo4j.index.internal.gbptree.GBPTree$SingleWriter.internalMerge(GBPTree.java:1708) ~[neo4j-index-4.4.11.jar:4.4.11]
	at org.neo4j.index.internal.gbptree.GBPTree$SingleWriter.merge(GBPTree.java:1695) ~[neo4j-index-4.4.11.jar:4.4.11]
	at org.neo4j.internal.id.indexed.IdRangeMarker.markDeleted(IdRangeMarker.java:165) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.internal.id.BufferingIdGenerator$1.markDeleted(BufferingIdGenerator.java:64) ~[neo4j-id-generator-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$ChangedIds.lambda$accept$6899bd4f$1(IdGeneratorUpdatesWorkSync.java:161) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.eclipse.collections.impl.list.mutable.primitive.LongArrayList.each(LongArrayList.java:588) ~[eclipse-collections-10.4.0.jar:?]
	at org.eclipse.collections.impl.list.mutable.primitive.LongArrayList.forEach(LongArrayList.java:577) ~[eclipse-collections-10.4.0.jar:?]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$ChangedIds.accept(IdGeneratorUpdatesWorkSync.java:151) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$IdGeneratorUpdateWork.apply(IdGeneratorUpdatesWorkSync.java:204) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$IdGeneratorUpdateWork.apply(IdGeneratorUpdatesWorkSync.java:177) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.util.concurrent.WorkSync.doSynchronizedWork(WorkSync.java:222) ~[neo4j-concurrent-4.4.11.jar:4.4.11]
	at org.neo4j.util.concurrent.WorkSync.tryDoWork(WorkSync.java:153) ~[neo4j-concurrent-4.4.11.jar:4.4.11]
	at org.neo4j.util.concurrent.WorkSync.applyAsync(WorkSync.java:120) ~[neo4j-concurrent-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$ChangedIds.applyAsync(IdGeneratorUpdatesWorkSync.java:168) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.applyInternal(IdGeneratorUpdatesWorkSync.java:122) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.apply(IdGeneratorUpdatesWorkSync.java:102) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.storageengine.util.IdGeneratorUpdatesWorkSync$Batch.close(IdGeneratorUpdatesWorkSync.java:129) ~[neo4j-storage-engine-util-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.close(IOUtils.java:152) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.close(IOUtils.java:188) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.io.IOUtils.closeAll(IOUtils.java:116) ~[neo4j-io-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.BatchContextImpl.close(BatchContextImpl.java:79) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	at org.neo4j.internal.recordstorage.RecordStorageEngine.apply(RecordStorageEngine.java:511) ~[neo4j-record-storage-engine-4.4.11.jar:4.4.11]
	... 52 more
2024-11-11 02:38:30.568+0000 INFO  [c.n.d.DbmsReconciler] Database 'neo4j' is requested to transition from STARTED{db=neo4j/6d79d176} to FAILED/QUARANTINED{db=neo4j/6d79d176} by Panic(priority):2
```

**Comment by arun-bedrock** [comment]
I also snapshotted the transaction logs before, including and after this crash, and the query.logs near this crash so I can help debug from those if you have commands you'd want me to run -- but I can't share those out.

**Comment by tinwelint** [comment]
Thanks, I'll have a look at those id generator logs!

**Comment by tinwelint** [comment]
I see the weird deletion of that specific ID in the logs you sent me. It's a record ID that is deleted from before, a new transaction re-allocates it and then when it commits that transaction it for some reason decides to delete that record. It surely looks like a bug, but it's hard to tell exactly what causes it from that log (it just says, I'm doing X to this ID, I'm doing Y to that ID a.s.o.)

I have some idea and can look closer if I can find some edge case that could somehow cause this.

**Comment by adam-bedrock** [comment]
Hey @tinwelint, I'm @arun-bedrock's coworker. Wondering if there is some sort of temporary workaround you can think of for this issue. Is there something we should be avoiding in our code to avoid this issue. Also, let us know if you need any other information from our side.

**Comment by tinwelint** [comment]
@adam-bedrock Hi, no I don't know exactly what's causing it a.t.m. so therefor I don't know a work-around unfortunately... other than moving to 5.25 and migrate to block format :) 

Are you using some `apoc` to run these write queries in parallel, or you're just firing these queries "plain and simple" over bolt? I'll have to find a way to try and reproduce your problem locally to understand this I think. Hopefully your initial info about the queries you're running will help at that.

**Comment by adam-bedrock** [comment]
I used to send queries in parallel, but now I'm just sending a single batch write at a time because I was worried that the parallelism might've been causing an issue. We started seeing a set of failures after running two write heavy flows together in parelel accidentally, wondering if the paralell jobs could've caused corruption in the database.

Why would updating neo4j fix this issue?

**Comment by arun-bedrock** [comment]
We just write with these batch queries with: 
`
UNWIND $items AS item
MATCH (n1:%s {key: item.key})
SET n1 += item.properties
`

OR for edge writing:
```
MATCH (n1:%s {key: item.fromKey})
MATCH (n2:%s {key: item.toKey})
MATCH (n1)-[r1:%s]->(n2)
SET r1 += item.properties
```

Our current theory is the unique constraint we have on our nodes for `key` is getting out of sync. The index is showing nodes as still existing but they're empty when you follow the index to the actual node (no `key` property or any other property exists on the node). Our queries fail when trying to use these nodes for writing and this potentially causes the crash. We find these nodes with `apoc.path.spanningTree`. We have tried dropping the index and re-adding it and that has fixed some of the issues but we still see some corruption (places where we can't write/update edges)

Our query for finding these nodes:
```
MATCH p=(n:%[2]s) <- [e:ENTITLED {%[1]s:"True"}] - (m)
		where n.key in $batch and (e.Derived%[1]s is null or e.Derived%[1]s <> "True")
		with collect(m) as startNodes, collect(m.key) as startKeys, collect(p) as toStartNodes,n
		with apoc.map.fromLists(startKeys, toStartNodes) as prefixs, startNodes,n
		call apoc.path.spanningTree(startNodes, {maxLevel: 10, relationshipFilter:'ASSUME<', labelFilter:'>User|>Role|>ManageGrants|>Anyone', bfs:true, beginSequenceAtStart:true}) yield path
		with path, n, nodes(path)[-1] as source, nodes(path)[0] as startNode,  prefixs
		with apoc.path.combine(prefixs[startNode.key], path) as path, n, source
		with reverse([x in nodes(path)|{key: x.key,name:x.name,types:labels(x),thirdParty: x.third_party = "True"}]) as displayPath, path, n, source, relationships(path)[0] as directEntitlement
		return
		source.key as from_key,
		labels(source) as from_types,
		n.key as to_key,
		labels(n)[0] as to_type,
		{
		Derived%[1]sPath: apoc.convert.toJson(displayPath),
		} as properties
```

**Comment by tinwelint** [comment]
> I used to send queries in parallel, but now I'm just sending a single batch write at a time because I was worried that the parallelism might've been causing an issue. We started seeing a set of failures after running two write heavy flows together in parelel accidentally, wondering if the paralell jobs could've caused corruption in the database.
> 
> Why would updating neo4j fix this issue?

Because of the newer database format (called `block`) that is very unlikely to have this problem. You'd have to explicitly migrate your db to it tho, even after upgrading to e.g. 5.25.

**Comment by tinwelint** [comment]
Oh, so you're seeing uniqueness constraint issues too? Don't think there are any known issues like that either. Very interesting.

**Comment by arun-bedrock** [comment]
Well we think that might be part of the the root cause for why the database is crashing. We have taken two steps so far: 
1) Dropping the uniqueness constraint and adding it back a few minutes later
2) recreating all edges (deleting and creating them again)

We're hoping after these changes the crashes won't happen anymore. We still aren't sure how the database got in this state though. I will update if the database crashes again after these changes.

**Comment by tinwelint** [comment]
@arun-bedrock great. And I will (whenever time permits) try to reproduce this issue, following your instructions as closely as I can.

**Comment by arun-bedrock** [comment]
Have you made any progress?

**Comment by tinwelint** [comment]
@arun-bedrock no. I've spent a little time on it, perhaps more soon. But haven't been able to reproduce yet.

**Comment by tinwelint** [comment]
@arun-bedrock I've tried various approaches but haven't been able to find a way to, regardless of input or access pattern, reproduce this.

So you say your write queries look like this: `UNWIND $items AS item MATCH (n1:%s {key: item.key}) SET n1 += item.properties` What's `$items` here? Not real data by some example, and where they come from. Can they contain null values?

**Comment by arun-bedrock** [comment]
yes they can contain null values. Thanks for continuing to investigate. Here is an anonymized example of items. 

```
{
  "items": [
    {
      "toKey": "DATABASE.SCHEMA:SCHEMA2.TABLE_NAME_WITH_PREFIX",
      "fromKey": "USER1@DOMAIN.COM",
      "properties": {
        "Write": "True",
        "ObtainedFromLocalUser": null,
        "WritePolicies": null,
        "ObtainedFromSnowflakeManagedSchemaOwnership": null,
        "DerivedWrite_Added": null,
        "ObtainedFromRedshift": null,
        "ObtainedFromIam": null,
        "ObtainedFromOkta": null,
        "update_time": 1728940947,
        "DerivedWrite": "True",
        "WriteActions": null,
        "ObtainedFromSnowflake": "True",
        "PartialWriteObjects": null,
        "PartialWrite": null,
        "FullWrite": null,
        "DerivedWritePath": "[{\"thirdParty\":null,\"types\":[\"User\",\"Principal\"],\"name\":\"USER1@DOMAIN.COM\",\"key\":\"USER1@DOMAIN.COM\"},{\"thirdParty\":null,\"types\":[\"Role\",\"Principal\"],\"name\":\"SYSTEM_ROLE1\",\"key\":\"snowflake:db:SYSTEM_ROLE1\"},{\"thirdParty\":null,\"types\":[\"Role\",\"Principal\"],\"name\":\"CUSTOM_ROLE1\",\"key\":\"snowflake:db:CUSTOM_ROLE1\"},{\"thirdParty\":null,\"types\":[\"Role\",\"Principal\"],\"name\":\"SCHEMA_WRITER_ROLE\",\"key\":\"snowflake:db:SCHEMA_WRITER_ROLE\"},{\"thirdParty\":null,\"types\":[\"Table\"],\"name\":\"DATABASE.SCHEMA:SCHEMA2.TABLE_NAME_WITH_PREFIX\",\"key\":\"DATABASE.SCHEMA:SCHEMA2.TABLE_NAME_WITH_PREFIX\"}]",
        "DerivedWrite_Added_time": null,
        "Write_update_time": 1728940947
      }
    },
    {
      "toKey": "DATABASE.SCHEMA:SCHEMA2.TABLE_NAME_WITH_PREFIX",
      "fromKey": "USER2@DOMAIN.COM",
      "properties": {
        "Write": "True",
        "ObtainedFromLocalUser": null,
        "WritePolicies": null,
        "ObtainedFromSnowflakeManagedSchemaOwnership": null,
        "DerivedWrite_Added": null,
        "ObtainedFromRedshift": null,
        "ObtainedFromIam": null,
        "ObtainedFromOkta": null,
        "update_time": 1728940947,
        "DerivedWrite": "True",
        "WriteActions": null,
        "ObtainedFromSnowflake": "True",
        "PartialWrite": null,
        "PartialWriteObjects": null,
        "FullWrite": null,
        "DerivedWritePath": "[{\"thirdParty\":null,\"types\":[\"User\",\"Principal\"],\"name\":\"USER2@DOMAIN.COM\",\"key\":\"USER2@DOMAIN.COM\"}]"
      }
    }
  ]
}
```

After upgrading to 5.0, we haven't experienced this issue by the way -- thanks for the suggestion to make the migration.
